### ISSUE TITLE
AddPath request to BGP daemon failed with code: 2 message not all flowspec component bytes available

### ISSUE THREAD

**Issue by pavel-odintsov** [osrg/gobgp#2292]
Hello!

Sorry about reporting many issues in same time. We're porting our very old GoBGP API to fresh version of GoBGP and appears that many things have changed.

In this case we're creating IPv4 Flow Spec announce using binary MP reach attributes. We create 3 attributes and we pass MP Reach attribute as NLRI in same time. 

For following announce:
```
We have generated BGP Flow Spec text representation for your announce: flow route source 11.22.33.44/32 destination 127.0.0.0/32 protocol [ tcp ] tcp-flags [ syn ] source-port [ =53 =5353 ] destination-port [ =80 ] packet-length [ =777 =1122 ] fragment [ is-fragment dont-fragment ] rate-limit 1024
```

We receive following error during attempt to announce it:
```
2020-05-31 22:49:41,220 [DEBUG] Announce 3 attributes
2020-05-31 22:49:41,221 [ERROR] AddPath request to BGP daemon failed with code: 2 message not all flowspec component bytes available
2020-05-31 22:49:41,221 [ERROR] Could not announce flow spec with GoBGP
```

I was able to track it in code https://github.com/osrg/gobgp/blob/833188f52610dcf47c57250788b36625b10a8925/pkg/packet/bgp/bgp.go#L4155 but I do not understand intentions of such change. 

I suppose it may be caused by our single byte encoding for TCP flags:
```
class __attribute__((__packed__)) bgp_flowspec_one_byte_byte_encoded_tcp_flags_t {
    public:
    // Just drop 8 bytes from bgp_flowspec_two_byte_encoded_tcp_flags
    uint8_t fin : 1, syn : 1, rst : 1, psh : 1, ack : 1, urg : 1, ece : 1, cwr : 1; 

    bgp_flowspec_one_byte_byte_encoded_tcp_flags_t() {
        memset(this, 0, sizeof(*this));
    }    

    std::string print() {
        std::stringstream buffer;

        buffer << "cwr: " << uint32_t(cwr) << " "
               << "ece: " << uint32_t(ece) << " "
               << "urg: " << uint32_t(urg) << " "
               << "ack: " << uint32_t(ack) << " "
               << "psh: " << uint32_t(psh) << " "
               << "rst: " << uint32_t(rst) << " "
               << "syn: " << uint32_t(syn) << " "
               << "fin: " << uint32_t(fin);

        return buffer.str();
    }    
};
```

But it's perfectly legit. It can be encoded using one or two bytes.

**Comment by fujita** [comment]
I'm not familiar with this but looks like the code expects that the first byte is type, then one or two bytes for the flags. It doesn't match with the on-wire format?

**Comment by pavel-odintsov** [comment]
Hello!

Nop, I use standard attribute encoding inside as MP Reach NLRI.

**Comment by imcom** [comment]
```
type FlowSpecComponentItem struct {
	Op    uint8  `json:"op"`
	Value uint64 `json:"value"`
}

```
Could it be missing the `op` field in your encoded message ? GoBGP looks for 1st byte - type, and then 2nd byte - op and finally a value which is your TCP flag ?

**Comment by pavel-odintsov** [comment]
Thank you for reply! I've reduced my flow spec query to very basic one:
```
set flowspec '{ "source_prefix": "4.0.0.0/24", "destination_prefix": "127.0.0.0/24", "action_type": "discard" }'
```

And it still fails with error.

I've added some debug tracing to distinguish between 5 possible "not all flowspec component bytes available" in code and I've identified path: https://github.com/osrg/gobgp/blob/v2.27.0/pkg/packet/bgp/bgp.go#L4418

**Comment by pavel-odintsov** [comment]
Hello again!

I think I've solved it but quite peculiar way.

- I skipped MP Reach NLRI and Flow Spec (which carries MP REACH NLRI attribute type, length and flags)
- I skipped MP Reach Flow spec header (which carries Address Family Identifier, Subsequent Address Family Identifier, Length of Next Hop Network Address (1 octet), Network Address of Next Hop (variable), Reserved (1 octet), Network Layer Reachability Information (variable)

I pushed structure encoded in following way:

- 1 byte length field
- All attributes encoding flow spec parameters (ports, lengths, addresses)

After that in attributes section I've pushed normally encoded MP Reach Flow Spec attribute, origin and rate as extended community.

It worked perfectly fine!

But I think more natural way will be empty NLRI for Flow spec and then all data encoded in attributes. 

Thank you so much for your assistance at official chat! :)
